pd/src/components/shielded_pool.rs

```diff
@@ -271,27 +269,48 @@ impl ShieldedPool {
             },
             source,
         )
-        .await;
+        .await?;
 
         Ok(())
     }
 
     #[instrument(skip(self, source, output_body))]
-    async fn add_note(&mut self, output_body: output::Body, source: NoteSource) {
+    async fn add_note(&mut self, output_body: output::Body, source: NoteSource) -> Result<()> {
         tracing::debug!(commitment = ?output_body.note_commitment, "appending to NCT in component");
         // 1. Insert it into the NCT
         self.note_commitment_tree
-            .append(&output_body.note_commitment);
+            .insert(penumbra_tct::Forget, output_body.note_commitment)?;
         // 2. Record its source in the JMT
         self.overlay
             .set_note_source(&output_body.note_commitment, source)
             .await;
         // 3. Finally, record it in the pending compact block.
         self.compact_block.outputs.push(output_body);
+        Ok(())
     }
 
     #[instrument(skip(self))]
     async fn write_block(&mut self) -> Result<()> {
+        // Determine if we are at an epoch boundary, and create a new epoch in the commitment tree
+        // if so; otherwise, just create a new block
+        let epoch_duration = self.overlay.get_epoch_duration().await?;
+        let height = self.compact_block.height;
+        if height != 0 {
+            if height % epoch_duration == 0 {
+                self.note_commitment_tree
+                    .insert_epoch(penumbra_tct::Epoch::new())?;
+            } else {
+                self.note_commitment_tree
+                    .insert_block(penumbra_tct::Block::new())?;
+            }
+        }
+
+        // Check that the position we're starting at matches the height of the block
+        assert_eq!(
+            u64::from(self.note_commitment_tree.position()) >> 16,
+            height
+        );
```

Comment:
This assertion is wrong, because it does not compute the height correctly (it will fail as soon as the first epoch ends, halting the chain).

Reply:
Fixed in https://github.com/penumbra-zone/penumbra/pull/617/commits/cda7931722303f4894e333fad8e1d29b65a65305